File: swift/problem744.swift
class Solution {
    func nextGreatestLetter(_ letters: [Character], _ target: Character) -> Character {
        var tail = letters.count - 1
        if (letters[tail] <= target) {
            return letters[0]
        }
        var head = 0; var mid = 0
        
        while (head < tail) {
            mid = (head+tail)/2
            if (letters[mid] <= target) {
                head = mid+1
            } else {
                tail = mid
            }
        }
        
        return letters[head]
    }
}

File: cpp/problem744.cpp
#include "mystd.h"
using namespace std;

/**
 * Perhaps the best but still most efficient way is linear searching, 
 * since maximally we can only have 26 characters in letters, and linear
 * search needs no multiplication and division, saving the time.
 */
class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
    	size_t i = 0;
    	while( i < letters.size() && letters[i] <= target) {
    	    ++i;
    	}
    	if( i == letters.size() ) return letters.front();
    	return letters[i];
    }
};

int main(int argc, char const *argv[]){
	// test example in Leetcode question
	vector<char> letters = {'c','f','j'};
	Solution sol;
	cout << sol.nextGreatestLetter(letters,'a') << endl;
	cout << sol.nextGreatestLetter(letters,'c') << endl;
	cout << sol.nextGreatestLetter(letters,'d') << endl;
	cout << sol.nextGreatestLetter(letters,'g') << endl;
	cout << sol.nextGreatestLetter(letters,'j') << endl;
	cout << sol.nextGreatestLetter(letters,'k') << endl;
	
	return 0;
}